### ISSUE TITLE
twisted.web.static.File doesn't support multiple byte ranges

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3574]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)| @mwhudson reported|
|-|-|
|Trac ID|trac#3574|
|Type|enhancement|
|Created|2008-12-11 21:55:57Z|
|Branch|https://github.com/twisted/twisted/tree/t.w.static.File-multiple-ranges-3574-2|

It grew support for single ranges recently, but it still doesn't support requests for more than one range at a time.

<details><summary>Searchable metadata</summary>

```
trac-id__3574 3574
type__enhancement enhancement
reporter__mwh mwh
priority__normal normal
milestone__ 
branch__branches_t_w_static_File_multiple_ranges_3574_2 branches/t.w.static.File-multiple-ranges-3574-2
branch_author__mwh mwh
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1229032557000000 1229032557000000
changetime__1246312861000000 1246312861000000
version__None None
owner__mwh mwh
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_1"></a>|@mwhudson commented|
|-|-|

(In [25635]) Branching to 't.w.static.File-multiple-ranges-3574'

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_2"></a>|@mwhudson commented|
|-|-|

(In [26805]) Branching to 't.w.static.File-multiple-ranges-3574-2'

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_3"></a>|@mwhudson commented|
|-|-|

(In [26807]) Branching to 't.w.static.File-multiple-ranges-3574-2'

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_4"></a>|@mwhudson removed owner|
|-|-|

OK, I think this is ready for review!

It's quite a chunky change, 1200 lines of diff.

The linked branch implements support for multipart responses for requests to static.File resource that specify multiply byte ranges.

The core is a new method, File.makeProducer.  This sets the headers in the request appropriately and returns IPullProducer that will produce the body of the request when you call start() on it.

Even though much of the logic is fairly generic, I've thoroughly resisted the temptation to reinvent streams as I did this :)  I think any generization can happen in a later branch.

A conceptual problem with this branch is that both the tests and the implementation follow my understanding of how all this works, so if there's a flaw in my understanding, the branch could be completely bogus :)  As a guard against this, I've tested it using it to serve bazaar branches over http, which is (a) the only http client I know of that actually requests multiple byte ranges and (b) my motivating use case.  It works.

While I was doing this I found and fixed [#3814](https://github.com/twisted/twisted/issues/3814).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

I'm going to give you some quick feedback, even though it's not a full review:
 
 * the `StaticProducer` classes are missing docstrings.
 * `test_mutipleRangeSetsContentHeaders` uses set, it should use twisted.compat.set
 * 
```
+        except ValueError, e:
+            log.msg("Ignoring malformed Range header %r" % (byteRange,))
```
 You're not using the "e" variable.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_6"></a>|@mwhudson commented|
|-|-|

All taken care of, thanks!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @mwhudson|
|-|-|

It looks really good, here are some stuff I spotted:

 * There is no test for the value of content-length in case of multiple ranges.
 * There is no test for the setting of content-encoding in _setContentHeaders
 * 
```
+        request = self.request
+        if done:
+            self.request = None
+        request.write(''.join(data))
+        if done:
+            request.unregisterProducer()
+            request.finish()
```

 I don't think there is a need to reset the request before writing the data. At least no test failed when I changed this code. The request.finish() call is not tested, too.

Do you have a sample web client that can be used to test this functionality?

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_8"></a>|@mwhudson removed owner|
|-|-|

I think I've addressed all these points, diff since review here: http://pastebin.ubuntu.com/199855/

> Do you have a sample web client that can be used to test this functionality?

bzr :)

Other than that, curl can make multiple range requests, but doesn't interpret the results.  I honestly don't know of any others.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_9"></a>|@mwhudson commented|
|-|-|

Oh, and I didn't say: thanks for the review!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to @mwhudson|
|-|-|

One last nitpicking:

```
+        raise NotImplementedError(self.resumeProducing)
+
+
+class NoRangeStaticProducer(StaticProducer):
```

One blank line missing.

Please merge!

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_11"></a>|@mwhudson set status to `closed`|
|-|-|

(In [27029]) Merge t.w.static.File-multiple-ranges-3574-2

Author: mwhudson
Reviewer: therve
Fixes: [#3574](https://github.com/twisted/twisted/issues/3574)
Fixes: [#3814](https://github.com/twisted/twisted/issues/3814)

Support multiple byteranges in the Range: header in requests for
static resources.

In the process fix a bug with Content-Length being wrong when the
request range only partially overlaps with the resource.
